pkg/orchestrator/scheduler/ops_job.go

```diff
@@ -0,0 +1,163 @@
+	"github.com/bacalhau-project/bacalhau/pkg/model"
+	"github.com/bacalhau-project/bacalhau/pkg/models"
+	"github.com/bacalhau-project/bacalhau/pkg/orchestrator"
+	"github.com/google/uuid"
+	"github.com/rs/zerolog/log"
+)
+
+// OpsJobScheduler is a scheduler for batch jobs that run until completion
+type OpsJobScheduler struct {
+	jobStore       jobstore.Store
+	planner        orchestrator.Planner
+	nodeDiscoverer orchestrator.NodeDiscoverer
+	nodeRanker     orchestrator.NodeRanker
+}
+
+type OpsJobSchedulerParams struct {
+	JobStore       jobstore.Store
+	Planner        orchestrator.Planner
+	NodeDiscoverer orchestrator.NodeDiscoverer
+	NodeRanker     orchestrator.NodeRanker
+}
+
+func NewOpsJobScheduler(params OpsJobSchedulerParams) *OpsJobScheduler {
+	return &OpsJobScheduler{
+		jobStore:       params.JobStore,
+		planner:        params.Planner,
+		nodeDiscoverer: params.NodeDiscoverer,
+		nodeRanker:     params.NodeRanker,
+	}
+}
+
+func (b *OpsJobScheduler) Process(ctx context.Context, evaluation *models.Evaluation) error {
+	ctx = log.Ctx(ctx).With().Str("JobID", evaluation.JobID).Str("EvalID", evaluation.ID).Logger().WithContext(ctx)
+
+	job, err := b.jobStore.GetJob(ctx, evaluation.JobID)
+	if err != nil {
+		return fmt.Errorf("failed to retrieve job %s: %w", evaluation.JobID, err)
+	}
+	// Retrieve the job state
+	jobState, err := b.jobStore.GetJobState(ctx, evaluation.JobID)
+	if err != nil {
+		return fmt.Errorf("failed to retrieve job state for job %s when evaluating %s: %w",
+			evaluation.JobID, evaluation, err)
+	}
+
+	// Plan to hold the actions to be taken
+	plan := models.NewPlan(evaluation, &job, jobState.Version)
+
+	existingExecs := execSetFromSliceOfValues(jobState.Executions)
+	nonTerminalExecs := existingExecs.filterNonTerminal()
+
+	// early exit if the job is stopped
+	if jobState.State.IsTerminal() {
+		nonTerminalExecs.markStopped(execNotNeeded, plan)
+		return b.planner.Process(ctx, plan)
+	}
+
+	// Retrieve the info for all the nodes that have executions for this job
+	nodeInfos, err := existingNodeInfos(ctx, b.nodeDiscoverer, nonTerminalExecs)
+	if err != nil {
+		return err
+	}
+
+	// Mark executions that are running on nodes that are not healthy as failed
+	nonTerminalExecs, lost := nonTerminalExecs.filterByNodeHealth(nodeInfos)
+	lost.markStopped(execLost, plan)
+
+	// create new executions only if this is the first time we are evaluating this job
+	var newExecs execSet
+	allFailed := existingExecs.filterFailed().union(lost)
+	if len(existingExecs) == 0 {
+		newExecs, err = b.createMissingExecs(ctx, &job, plan)
+		if err != nil {
+			b.handleFailure(nonTerminalExecs, allFailed, plan, err)
+			return b.planner.Process(ctx, plan)
+		}
+	}
+
+	// mark job as completed if there are no more active or new executions
+	if len(nonTerminalExecs) == 0 && len(newExecs) == 0 {
+		if len(allFailed) > 0 {
+			b.handleFailure(nonTerminalExecs, allFailed, plan, errors.New(""))
```

Comment:
should `errors.New("")` contain a helpful message?

Reply:
`handleFailure` will use the failure message from the latest execution, and fall back to the passed error if no failed execution was found, which won't be the case here